core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -54,41 +87,28 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
 
   private[transaction]
   def drainQueuedTransactionMarkers(txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]): Iterable[RequestAndCompletionHandler] = {
-    brokerStateMap.flatMap {case (brokerId: Int, destAndMarkerQueue: DestinationBrokerAndQueuedMarkers) =>
-      val markersToSend: java.util.List[CoordinatorEpochAndMarkers] = new util.ArrayList[CoordinatorEpochAndMarkers] ()
-      destAndMarkerQueue.markersQueue.drainTo (markersToSend)
-      markersToSend.groupBy{ epochAndMarker => (epochAndMarker.metadataPartition, epochAndMarker.coordinatorEpoch) }
-        .map { case((metadataPartition: Int, coordinatorEpoch:Int), buffer: mutable.Buffer[CoordinatorEpochAndMarkers]) =>
-          val txnMarkerEntries = buffer.flatMap{_.txnMarkerEntries }.asJava
-          val requestCompletionHandler = new TransactionMarkerRequestCompletionHandler(
-            this,
-            txnMarkerPurgatory,
-            CoordinatorEpochAndMarkers(metadataPartition, coordinatorEpoch, txnMarkerEntries),
-            brokerId)
-          RequestAndCompletionHandler(destAndMarkerQueue.destBrokerNode, new WriteTxnMarkersRequest.Builder(coordinatorEpoch, txnMarkerEntries), requestCompletionHandler)
-        }
+    brokerStateMap.flatMap {case (brokerId: Int, brokerRequestQueue: BrokerRequestQueue) =>
+      brokerRequestQueue.eachMetadataPartition{ case(partitionId:Int, queue:BlockingQueue[TxnMarkerEntry]) =>
+        val markersToSend: java.util.List[TxnMarkerEntry] = new util.ArrayList[TxnMarkerEntry]()
+        queue.drainTo(markersToSend)
+        val requestCompletionHandler = new TransactionMarkerRequestCompletionHandler(this, txnMarkerPurgatory, partitionId, markersToSend, brokerId)
+        RequestAndCompletionHandler(brokerRequestQueue.node, new WriteTxnMarkersRequest.Builder(markersToSend), requestCompletionHandler)
+      }
     }
   }
 
 
   def addOrUpdateBroker(broker: Node) {
-    if (brokerStateMap.contains(broker.id())) {
-      val brokerQueue = brokerStateMap(broker.id())
-      if (!brokerQueue.destBrokerNode.equals(broker)) {
-        brokerStateMap.put(broker.id(), DestinationBrokerAndQueuedMarkers(broker, brokerQueue.markersQueue))
-        trace(s"Updated destination broker for ${broker.id} from: ${brokerQueue.destBrokerNode} to: $broker")
-      }
-    } else {
-      val markersQueue = new LinkedBlockingQueue[CoordinatorEpochAndMarkers]()
-      brokerStateMap.put(broker.id, DestinationBrokerAndQueuedMarkers(broker, markersQueue))
-      trace(s"Added destination broker ${broker.id}: $broker")
+    brokerStateMap.putIfAbsent(broker.id(), new BrokerRequestQueue(broker)) match {
+      case Some(brokerQueue) => brokerQueue.maybeUpdateNode(broker)
```

Comment:
@guozhangwang, I was just saying that semantically it's better to use the broker id as the key because `Node`'s hashCode and equality are based on various fields. It also avoids the scenario that Damian pointed out, which makes it a bit fragile. Even if it may work today.